### ISSUE TITLE
FLIR THermal Dataset for ADAS

### ISSUE THREAD

**Issue by abhimanyuiris96** [ultralytics/yolov5#6301]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am trying to finetune yolov5x to the FLIR thermal dataset for ADAS and while the training starts out good, it gets stuck halfway through and then shows no improvement at all. Any suggestions for hyperparameter tuning or configurations that I could try?

Link to the dataset:-
https://www.kaggle.com/deepnewbie/flir-thermal-images-dataset

**Comment by glenn-jocher** [comment]
@abhimanyuiris96 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. The below recommendations should work well for your FLIR dataset.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by iceisfun** [comment]
I was able to train it fine using yolov5s batch size 30 epochs 20

```
     Epoch   gpu_mem       box       obj       cls    labels  img_size
     19/19     6.11G   0.03771   0.03979  0.003014       110       640: 100%|██████████| 235/235 [00:27<00:00,  8.51it/s]                                                           
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 41/41 [00:02<00:00, 14.15it/s]                                             
                 all        814       7102      0.854      0.769      0.849      0.451

```

**Comment by rdabane** [comment]
> I was able to train it fine using yolov5s batch size 30 epochs 20
> 
> ```
>      Epoch   gpu_mem       box       obj       cls    labels  img_size
>      19/19     6.11G   0.03771   0.03979  0.003014       110       640: 100%|██████████| 235/235 [00:27<00:00,  8.51it/s]                                                           
>                Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 41/41 [00:02<00:00, 14.15it/s]                                             
>                  all        814       7102      0.854      0.769      0.849      0.451
> ```

@iceisfun ,
Could you please paste here the details of fine tuning here .. especially lr etc .. or the command you used ..
Thanks.

**Comment by glenn-jocher** [comment]
Hello @rdabane,
It seems like @iceisfun was able to train their model fine using yolov5s batch size 30 epochs 20. Regarding the details of fine-tuning, we would need more information from @iceisfun such as the command used and learning rate (lr) to provide more insight. Could you please wait for their response? Thank you.

**Comment by rdabane** [comment]
@glenn-jocher ,
Thanks for your quick update. I'll look out for the @iceisfun response.
Thanks.

**Comment by glenn-jocher** [comment]
Hello @rdabane,

Thank you for your prompt response. We will await @iceisfun's reply regarding the details of their fine-tuning process, in order to provide more information.

If you have any other questions or concerns, please feel free to ask. We're here to help.

Best regards,

**Comment by elmonkey** [comment]
Hi. 

I was able to finetune the yolov5 object detector using flir's thermal dataset. Hopefully, this blurb can save people some time. 

First, the `trick is having the dataset in the right YOLOv5 format` (annotations) and directory structure. 

The high level description of what I did:

1. get the dataset from: https://adas-dataset-v2.flirconservator.com/#downloadguide

2. convert the dataset from COCO to YOLOv5 (I used [globox](https://github.com/laclouis5/globox) --- there might be other methods like roboflow - idk)
   * run on train, validation, and test task. I like having <task>/images/*.jpg and <task>/labels/*.txt, else you have to make other unnecessary changes to YOLO to make sure things work well (where to find the annotations)

```
# from globox
from globox import AnnotationSet

# load annotations
coco = AnnotationSet.from_coco(file_path=input_path) 
# display stats
coco.show_stats()

# SAVE TO YOLOv5
coco.save_yolo_v5(
    save_dir = <output_path>,
    label_to_id={
        "person": 0,
        "car": 1,
        ... etc
    }
)

```
  * NOTE: requires a bit of manual work organizing folders, but it does work  (I've tried a few that didn't) -- I modified to omit certain classes. 
 
The `data.yaml`:
```
train: $YOLOv5/images_thermal_train/images
val: $YOLOv5/images_thermal_val/images
test: $YOLOv5/video_thermal_test/images
nc: <N>
names: 
    0: 'person'
    1:  'car'
    ...  etc
```
remember to correctly set the `data.yaml` file
    
3. Modified ultralytics yolov5 notebook to fine-tune yolov5L-pretrained weights using batchsize=32, imsize=480, and I get nice detections after 10 epochs. You will need to modify and/or create a yolov5/data/hyps/<some_name>.yaml file with things like:
```
...

lr0: 0.012
lrf: 0.115
...
etc.
```

4. Inspect the input data and results. I like using supervision detection datasets (by the roboflow ppl), but the latest ultralytics pkg has nice visualization (usually saved under the `runs` folder)

Good luck,

**Comment by glenn-jocher** [comment]
@elmonkey hi,

Thank you for sharing your experience and providing some helpful tips for fine-tuning YOLOv5 with FLIR's thermal dataset. It's great to hear that you were able to successfully convert the dataset from COCO to YOLOv5 format using globox. It seems like a useful tool for converting annotations.

Organizing the dataset with the correct directory structure and creating the `data.yaml` file with the appropriate paths and class names are crucial steps, and it's good that you mentioned them.

Modifying the ultralytics yolov5 notebook and creating a hyps file for setting hyperparameters is also important to achieve good results during fine-tuning. It's nice that you were able to obtain satisfactory detections after ten epochs with a batch size of 32 and image size of 480.

Inspecting the input data and results is essential for evaluating the performance of the model, and it's helpful that you mentioned utilizing supervision detection datasets and the visualization capabilities in the latest ultralytics package.

Thank you once again for sharing your insights and tips. Good luck with your further experiments!
